core/ir/synthetic/encoding_common.h

```diff
@@ -0,0 +1,147 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _SYNTHETIC_ENCODING_COMMON_H_
+#define _SYNTHETIC_ENCODING_COMMON_H_
+
+/**
+ * This synthetic ISA (which we call DR_ISA_REGDEPS) is a made up ISA that has the purpose
+ * of preserving register dependencies and giving hints on the type of operations each
+ * instruction is performing.
+ * For this reason the majority of operations that would normally work on instructions
+ * coming from an actual ISA (e.g., DR_ISA_AMD64) are not supported.
+ * We support instr_convert_to_isa_regdeps(), which converts an #instr_t of an actual ISA
+ * to a DR_ISA_REGDEPS instruction.
+ * Other operations we support (related to encoding and decoding of a DR_ISA_REGDEPS
+ * #instr_t) are: instr_encode(), instr_encode_to_copy(), decode(), decode_common().
+ * DR_ISA_REGDEPS #instr_t can only return the encoded information described in the
+ * encoding scheme (core/ir/synthetic/encoding_common.h): categories (from #dr_category_t)
+ * to indicate the type of operation performed (e.g., a load, a math operation, floating
+ * point math operation, a branch, etc.), arithmetic flags, number of source and
+ * destination register operands, register operands and their size (which is the same for
+ * all operands), instruction length, and ISA mode (which is DR_ISA_REGDEPS).
+ * Querying additional information (e.g., the instruction opcode) will return the default
+ * value generated by instr_create() (i.e., zero).
+ */
+
+/* XXX i#1684: Note that this is part of a larger issue, where lack of cross-arch support
+ * in the same build of DynamoRIO is limiting us.
+ */
+
+/* Here we describe the encoding scheme for the Synthetic ISA that is enforced in decode.c
+ * and encode.c.
+ *
+ * Encoded instructions are 4 byte aligned.
+ *
+ * All instruction encodings begin with the following 4 header bytes, which follow this
+ * scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *        category         eflags #src   #dst
+ *
+ * 22 bits, category: it's a high level representation of the opcode of an instruction.
+ * Each bit represents one category following #dr_instr_category_t.
+ * Note that an instruction can belong to more than one category, hence multiple bits can
+ * be set.
+ * 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at least
```

Comment:
Similar comment as above: Add bullets or indentation or blank lines so it's easier to see the separations here; I find it difficult to read as-is with this all-flush-left style here.

Reply:
Done. https://github.com/DynamoRIO/dynamorio/pull/6691#discussion_r1554658906